src/webui/api/torrentscontroller.cpp

```diff
@@ -1622,3 +1792,246 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::fetchMetadataAction()
+{
+    requireParams({u"source"_s});
+
+    const QString sourceParam = params()[u"source"_s].trimmed();
+    // must provide some value to parse
+    if (sourceParam.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    const QUrl sourceUrl = QUrl::fromPercentEncoding(sourceParam.toLatin1());
+    const QString source = sourceUrl.toString();
+
+    // metadata has already been requested
+    if (const auto iter = m_torrentSource.constFind(source); iter != m_torrentSource.constEnd())
+    {
+        const BitTorrent::InfoHash infoHash = iter.value();
+        // metadata has already been downloaded
+        if (isMetadataDownloaded(infoHash))
+        {
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+        }
+        else
+        {
+            setStatus(APIStatus::Async);
+            setResult(serializeInfoHash(infoHash));
+        }
+
+        return;
+    }
+
+    // http(s) url
+    if (Net::DownloadManager::hasSupportedScheme(source))
+    {
+        qDebug("Fetching torrent %s", qUtf8Printable(source));
+        const auto *pref = Preferences::instance();
+
+        Net::DownloadManager::instance()->download(Net::DownloadRequest(source).limit(pref->getTorrentFileSizeLimit())
+                , pref->useProxyForGeneralPurposes(), this, &TorrentsController::onDownloadFinished);
+
+        m_torrentSource.insert(source, {});
+
+        setStatus(APIStatus::Async);
+        setResult(QJsonObject {});
+        return;
+    }
+
+    // magnet link or info hash
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+        auto *session = BitTorrent::Session::instance();
+        // torrent already exists in transfer list
+        if (const BitTorrent::Torrent *const torrent = session->findTorrent(infoHash); torrent)
+        {
+            // torrent's metadata may not have downloaded yet
+            if (torrent->info().isValid())
+            {
+                setResult(serializeTorrentInfo(*torrent));
+            }
+            else
+            {
+                setStatus(APIStatus::Async);
+                setResult(serializeInfoHash(infoHash));
+            }
+
+            return;
+        }
+
+
+        // metadata may have already been downloaded under a different source URI
+        if (isMetadataDownloaded(infoHash))
+        {
+            // make torrent available for download (via /add API) using this
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
```

Comment:
It's already there!

Reply:
The key already exists in `m_torrentMetadata`, but it potentially has a different value. We update the value so that this info hash always stores the more recently submitted `TorrentDescriptor`.